compiler/parser/RdrHsSyn.hs

```diff
@@ -2781,6 +2858,35 @@ mkTemplateDecls lname@(L nloc name) fields (L _ decls) = do
       }
     pureUnit = mkApp (mkUnqualVar $ mkVarOcc "pure") (noLoc $ ExplicitTuple noExt [] Boxed)
 
+-- Generate `newtype` and `instance` declarations corresponding to a
+-- `template instance InstanceName = T Arg1 .. ArgN`.
+mkTemplateInstance
+  :: Located RdrName              -- ^ Name given to template instance
+  -> LHsType GhcPs                -- ^ Application of generic template to type arguments
+  -> P (OrdList (LHsDecl GhcPs))  -- ^ Resulting declarations (`newtype` and `instance` of `TInstance` class)
+mkTemplateInstance instName@(L instLoc _) templateApp
+  | (templateType, tyArgs) <- splitHsAppTysPs templateApp
+  , L _ (HsTyVar NoExt NotPromoted templateName) <- templateType = do
+      let templateString = occNameString $ rdrNameOcc $ unLoc templateName
+          tInstanceClass = mkUnqualClass $ mkInstanceClassName templateString
+          instType = unLoc $ mkHsAppTys tInstanceClass tyArgs
+          instDataName = mkRdrUnqual . mkDataOcc . occNameString . rdrNameOcc <$> instName
+          newTypeCon = L instLoc $ mkConDeclH98 instDataName Nothing Nothing $ PrefixCon [templateApp]
+      newTypeDecl <- mkTyData instLoc NewType Nothing (noLoc (Nothing, rdrNameToType instName)) Nothing [newTypeCon] (noLoc [])
+      return $ toOL [TyClD noExt <$> newTypeDecl, instDecl $ classInstDecl instType emptyBag]
+  | otherwise = addFatalError instLoc $ text $
+      occNameString (rdrNameOcc (unLoc instName)) ++ " is not an application of a generic template"
```

Comment:
This should make your TODO obsolete:
```suggestion
      getOccString instName ++ " is not an application of a generic template"
```

Reply:
`getOccString` doesn't work because there is no `NamedThing` instance for `RdrName` - it seems like there should be but maybe it's intentionally omitted for some reason. Anyway I'm adding a helper function for now.